[#49932] Convert file storages index page to primer design

akabiru opened the pull request:
#### https://community.openproject.org/projects/openproject/work_packages/49932

- [x] Adopt [OpenProject::PageHeader](https://github.com/opf/primer_view_components/pull/22) for page heading
   - [x] Confirm whether breadcrumbs remain unchanged
- [x] File storages "table" as [BorderBox](https://primer.style/design/components/border-box/rails/beta)
  - [x] Empty state
  - Possibility to edit/delete a storage ([Separate WP](https://community.openproject.org/work_packages/50212/relations))
- [x] Check on mobile 

![Screenshot 2023-10-03 at 3 15 10 PM](https://github.com/opf/openproject/assets/17295175/5a92a9ca-2bf7-42a8-9406-e6aad1ab4fad)

![Screenshot 2023-10-03 at 2 51 26 PM](https://github.com/opf/openproject/assets/17295175/39603e07-5962-4944-97ba-811200bdbbfb)



<details>
<summary>Mobile Screenshots 📱 </summary>

![Screenshot 2023-10-03 at 3 15 27 PM](https://github.com/opf/openproject/assets/17295175/0b761a33-38f8-4ca6-8845-a2a34df599d1)

![Screenshot 2023-10-03 at 2 51 42 PM](https://github.com/opf/openproject/assets/17295175/d973193f-42be-4d4d-a8f5-b88ed0e675f1)



</details>

akabiru committed:
feat[#49841 #49932]: Convert storages admin heading to `OpenProject::PageHeader`

akabiru committed:
UI[#49841 #49932]: Amend typography to get closer to design

HDinger commented on modules/storages/app/views/storages/admin/storages/index.html.erb:
```diff
@@ -1,31 +1,48 @@
-      <span class="button--text"><%= ::Storages::Storage.model_name.human %></span>
+<% if OpenProject::FeatureDecisions.storage_primer_design_active? %>
+  <%= render(Primer::OpenProject::PageHeader.new) do |component| %>
+    <% component.with_title(classes: "f1-light") do %>
```
Hi @akabiru 
after longer discussion with the product team about the desired font size (and font weight), we agreed on 24px as default for the PageHeader Component. I just changed that in our viewComponents fork. Thus, the extra class here is not needed any more as soon you rebase.

akabiru replied:
Thank you @HDinger , I'll rebase and update! 👍🏾

akabiru committed:
fix[#49932]: Adopt updated styling for the PageHeader

See https://github.com/opf/openproject/pull/13660/files#r1319791460

akabiru committed:
fix[#49932]: Update button styling to have + icon

akabiru committed:
fix[#49932]: Toggle breadcrumbs based on the feature flag

akabiru committed:
fix[#49932]: Refine storage button as link with aria & title

akabiru committed:
chore[#49932]: define component base for easier view helper integrations

akabiru committed:
fix[#49932]: Use slot component

akabiru committed:
fix[#49932]: Do not attempt to override primer base templates and components

akabiru committed:
fix[#49932]: Structure HTML, split by divs for upcoming styling

akabiru committed:
UI[#49932]: Add row id as storage dom ID

akabiru committed:
UI[#49932]: Use primer base component instead of rails view content tag

akabiru committed:
fix[#49932]: Extend ApplicationComponent, compose Primer components within

akabiru commented on modules/storages/app/components/storages/admin/storage_list_component.rb:
```diff
@@ -0,0 +1,118 @@
+  end
+
+  class BorderBoxComponent < Primer::Beta::BorderBox
+    def before_render
+      super
+
+      # Remove the _base.sass margin-left from <ul> tag
+      @list_arguments[:classes] = "ml-0"
+    end
```
Overriding the default list arguments- that are not overrideable via [the constructor](https://github.com/primer/view_components/blob/main/app/components/primer/beta/border_box.rb#L85) afaics. cc @HDinger

cbliard replied:
This is necessary because `frontend/src/global_styles/openproject/_base.sass` defines this at the end of the file:

```
ul,
ol
  margin-left: 2rem
```

It works fine with a plain `Primer::Beta::BorderBox` if these styles are removed. What would be the side-effects if these lines were removed?

HDinger replied:
The lines in `frontend/src/global_styles/openproject/_base.sass` are necessary for now to intend our list. Without them, all of our list would show their bullet points too far on the left (maybe even out of view). So we need that. I am aware, that inside the Primer::Beta::BorderBox it looks weird if we do not override it.
However, I am reyling on [this PR](https://github.com/opf/openproject/pull/13527) as it will add a general rule for [lists inside BorderBoxes](https://github.com/opf/openproject/pull/13527/files#diff-ebc8270818b7f26cb6ecc2c0c6b4752069453d5b02a91f6f08c5738ac8387c3cR34). We have to wait for that PR anyway to benefit from the flexbox helpers.

akabiru committed:
feat[#49841 #49932]: Define NewStorageButtonComponent

cbliard commented on modules/storages/app/components/storages/admin/new_storage_button_component.rb:
```diff
@@ -0,0 +1,58 @@
+      { scheme:,
+        size:, tag:,
+        href: new_admin_settings_storage_path,
+        aria: { label: },
+        title: label }
```
This will not be helpful from a screen reader perspective. The text being read with Orca is "Storage (pause) visited link".
[The title element should always be avoided](https://www.a11yproject.com/posts/title-attributes/), except for iframe tag. It is not accessible.
The aria-label is useless if it is the same as the text of the element.

To improve it from an accessibility stand point:
* if it acts as a button, use `role="button"`
* the `+` icon is invisible to the user. A sighted user might understand that it is a metaphor for "Add", but blind people will not. They will only "see" a "Storage" button without knowing the action
  * You can add `aria-label="Add new storage"` (for instance)
  * Changing the label from "Storage" to "Add new storage" would maybe make it clearer for some sighted users. That's something to discuss with the UX people.

With these modifications, Orca now reads it as "Add new storage (pause) push button".

akabiru replied:
This is super helpful Christophe, thank you 👍🏾

HDinger committed:
Extract template logic to .erb files and use helper for correct pluralisation and datetime format

cbliard commented on modules/storages/app/components/storages/admin/storage_list_component.rb:
```diff
@@ -0,0 +1,72 @@
+    end
+
+    def formatted_datetime(storage)
+      span_tag(" #{I18n.t('activity.item.created_on', datetime: helpers.format_time(storage.created_at.to_fs(:long)))}")
```
It should be `helpers.format_time(storage.created_at))` to properly format with time zones.

akabiru replied:
Good shout 👍🏾

cbliard commented:
I tried to extract the html rendering into erb files too, but with more html than ruby. I find having methods rendering html being called here and there harder to read. Having everything in one erb template seems simpler.

The modifications are in the branch implementation/49932-convert-the-index-page-to-primer-design--with-erb-template. Differences can be seen with this link https://github.com/opf/openproject/compare/implementation/49932-convert-the-index-page-to-primer-design...implementation/49932-convert-the-index-page-to-primer-design--with-erb-template although I did not find how to compare a tip of a branch to the tip of another branch.

Thoughts?

akabiru commented:
> I tried to extract the html rendering into erb files too, but with more html than ruby. I find having methods rendering html being called here and there harder to read. Having everything in one erb template seems simpler.
> 
> The modifications are in the branch implementation/49932-convert-the-index-page-to-primer-design--with-erb-template. Differences can be seen with this link [implementation/49932-convert-the-index-page-to-primer-design...implementation/49932-convert-the-index-page-to-primer-design--with-erb-template](https://github.com/opf/openproject/compare/implementation/49932-convert-the-index-page-to-primer-design...implementation/49932-convert-the-index-page-to-primer-design--with-erb-template) although I did not find how to compare a tip of a branch to the tip of another branch.
> 
> Thoughts?



> I tried to extract the html rendering into erb files too, but with more html than ruby. I find having methods rendering html being called here and there harder to read. Having everything in one erb template seems simpler.
> 
> The modifications are in the branch implementation/49932-convert-the-index-page-to-primer-design--with-erb-template. Differences can be seen with this link [implementation/49932-convert-the-index-page-to-primer-design...implementation/49932-convert-the-index-page-to-primer-design--with-erb-template](https://github.com/opf/openproject/compare/implementation/49932-convert-the-index-page-to-primer-design...implementation/49932-convert-the-index-page-to-primer-design--with-erb-template) although I did not find how to compare a tip of a branch to the tip of another branch.
> 
> Thoughts?

Cheers @cbliard , responded on Element- one thing that's missing is the `ul` margin left workaround https://github.com/opf/openproject/pull/13660#discussion_r1324388283 unless there's a different approach?

cbliard commented:
> Cheers @cbliard , responded on Element- one thing that's missing is the `ul` margin left workaround [#13660 (comment)](https://github.com/opf/openproject/pull/13660#discussion_r1324388283) unless there's a different approach?

As Henriette dropped it in [her last commit](https://github.com/opf/openproject/pull/13660/commits/d0a9410a497bea1302e3e63537b8187e432ea25d#diff-d890fe219e1f4dcb4aa4299a43c1c01b51553958dadc108191b688404549765fL109-L117), I chose to ignore it as well.

She said: 
> However, I am reyling on [this PR](https://github.com/opf/openproject/pull/13527) as it will add a general rule for [lists inside BorderBoxes](https://github.com/opf/openproject/pull/13527/files#diff-ebc8270818b7f26cb6ecc2c0c6b4752069453d5b02a91f6f08c5738ac8387c3cR34). We have to wait for that PR anyway to benefit from the flexbox helpers.

oliverguenther committed:
Fix form submissions when mixing primer with named submit buttons

This required us to use a timeout to re-enable a disabled submitter.
In turn, a few specs failed which now no longer did a form submit directly.

HDinger committed:
Style Box content according to visuals

akabiru committed:
Merge branch 'dev' into implementation/49932-convert-the-index-page-to-primer-design

akabiru committed:
feat[#49932]: Add blank slate

akabiru committed:
fix[#49932]: Use pluralize view helper

akabiru committed:
Revert "fix[#49932]: Use pluralize view helper"

This reverts commit 3bc87e37a645874f82d083ba46e7c3d154075917.

akabiru committed:
temp[#49932]: add header action buttones, pending design mockups

HDinger committed:
Merge remote-tracking branch 'origin/dev' into implementation/49932-convert-the-index-page-to-primer-design

HDinger committed:
Use grid layout for storage index page

akabiru committed:
chore[#49932]: remove provider & creator filters; out of scope

akabiru committed:
chore[#49932]: Switch from storages count to table headers

akabiru committed:
chore[#49932]: Define storage name as a link

akabiru committed:
chore[#49932]: remote host position to start, current default

akabiru committed:
chore[#49932]: Remote mute on link

akabiru committed:
chore[#49932]: Use font_weigth sys arg instead

akabiru committed:
UI[#49932]: update row items to text light

akabiru committed:
chore[#49932]: rubocop ammends

akabiru committed:
tests[#49932]: Add feature specs for index page

akabiru commented on spec/support/pages/messages/create.rb:
```diff
@@ -44,7 +44,7 @@ def add_text(text)
     end
 
-    def click_save
+    def click_save(expect_success: true)
```
@HDinger @oliverguenther I'm not sure I understand this change https://github.com/opf/openproject/pull/13660/commits/84cb848e418a7e599aa531388b17822581375a42 as it doesn't seem to change any behaviour 🤔

akabiru replied:
ping on this one @HDinger 😄

oliverguenther replied:
you can remove it, it was part of a temporary fix and I seem to have overlooked that part

akabiru committed:
fix[#49932]: Reusue the same storage button

akabiru committed:
chore[#49932]: Use multiline grid area definitions

To improve the readability of this code, it might be helpful to split the grid-template-columns property into multiple lines, with each value on its own line. This would make it easier to see the individual columns and their widths.

akabiru committed:
UI[#49932]: update to new design

akabiru committed:
UI[#49932]: Hide creator and timestamp headers titles in mobile

akabiru committed:
UI[#49932]: hide border-bottom on mobile screen

akabiru committed:
UI[#49932]: Resize primary buttons based on screen size

akabiru committed:
fix[#49932]: update tests to account for label in place of icon

akabiru committed:
fix[#49932]: update timestamp format

akabiru committed:
Merge branch 'dev' into implementation/49932-convert-the-index-page-to-primer-design

HDinger commented on config/locales/en.yml:
```diff
@@ -1318,6 +1318,8 @@ en:
         one: "1 second"
         other: "%{count} seconds"
+    formats:
+      long: "%Y-%m-%d %H:%M:%S %Z"
```
We already have a time format defined. So please double check with the product team whether this format is really required, or whether our existing one can be used.

akabiru replied:
Good shout, I'll check with Marc 👍🏾

akabiru replied:
Marc- revert back to activity timestamp

HDinger commented on frontend/src/stimulus/controllers/dynamic/buttons/op-primer-responsive-button.controller.ts:
```diff
@@ -0,0 +1,60 @@
+
+import { Controller } from '@hotwired/stimulus';
+
+export default class OpPrimerResponsiveButtonController extends Controller {
```
I'd really really like to avoid that class. First, it is super hard to maintain. You are applying primer classes and breakpoints in here which will not be updated if Primer changes them. Second, I'd question the need of it completely. Is it really necessary to switch to a smaller button on mobile?

akabiru replied:
Yeah, working this out with CSS was rather difficult as we have to btn switch classes- however I do agree that we can get rid of this if it's not a must have- will check with product 👍🏾

akabiru replied:
Marc- we can remove this

HDinger commented on modules/storages/app/components/storages/admin/storage_list_component.html.erb:
```diff
@@ -0,0 +1,78 @@
+      component.with_description { I18n.t('storages.storage_list_blank_slate.description') }
+
+      component.with_primary_action(
+        href: new_admin_settings_storage_path,
```
Are these always the same `button_options` as in the `NewStorageButtonComponent`? If yes, we should reuse them.

akabiru replied:
They are, although It seemed rather hacky to try and reuse them as primer primary_action slot is tightly coupled- Ideally it would have been better to just pass the `NewStorageButtonComponent`.

We could however do sth like

```ruby
component.with_primar_action(**NewStorageButtonComponent.new.button_options) do
  button.with_leading_visual_icon(icon: :plus)
  I18n.t("storages.label_storage")
end

```

HDinger commented on modules/storages/app/components/storages/admin/storage_list_component.html.erb:
```diff
@@ -0,0 +1,78 @@
+            end
+
+            unless storage.configured?
+              grid.with_area(:label, tag: :div) do
```
`label` is a bad name for a grid-area, as it is not not really explanatory. Something like `error-label` would be better

akabiru replied:
Ah, yeah- good point

HDinger commented on modules/storages/app/components/storages/admin/storage_list_component.sass:
```diff
@@ -0,0 +1,27 @@
+  grid-template-areas: "name label user provider time" "host host user provider time"
+
+  .op-principal--name
+    @extend .text-light
```
Colors should be set via the system_arguments of a component because it is less error-prone and we will notice when something changes on the Primer side

akabiru replied:
AFAICT, the `User::AvatarComponent` is not a primer component- we're still reusing the legacy avatar helper 🤔 

https://github.com/opf/openproject/blob/dev/app/components/users/avatar_component.rb

HDinger commented on modules/storages/app/components/storages/admin/storage_list_component.sass:
```diff
@@ -0,0 +1,27 @@
+    grid-row-gap: 5px
+
+    &--provider
+      display: none
```
There is a system argument, which you can set to hide elements on certain breakpoints `hide: :sm`. Same for the user and time element

akabiru replied:
Oh, I didn't know that- nice one 👍🏾

HDinger commented on modules/storages/app/controllers/storages/admin/storages_controller.rb:
```diff
@@ -166,7 +166,7 @@ def default_breadcrumb
   # Defines whether to show breadcrumbs on the page or not.
   def show_local_breadcrumb
-    true
+    !OpenProject::FeatureDecisions.storage_primer_design_active?
```
I am wondering whether we really need the feature flag after all. 🤔 Maybe we can discuss that later.

akabiru replied:
Since we've moved the edit & delete buttons into the storage edit/view page- removing the feature flag at this stage would be disruptive to the user as they wouldn't have a delete action 😬 My plan was to retain the flag and remove it once we have the edit page built out-

HDinger commented on modules/storages/spec/features/admin_storages_spec.rb:
```diff
@@ -31,12 +31,65 @@
+      it 'renders a list of storages' do
+        visit admin_settings_storages_path
+
+        expect(page).to have_css('.op-storage-list--name', text: complete_storage.name)
```
We have the patter to not test on classes any more as you'd have to adapt all tests if you change a class name. Instead we have a data attribute called `data-test-selector` which you set on those elements that you need, and then test for it. This selector is independent of the class and idealy not touched any more in the future.

HDinger commented on modules/storages/app/components/storages/admin/storage_list_component.sass:
```diff
@@ -0,0 +1,27 @@
+.op-storage-list
+  display: grid
+  grid-template-columns: fit-content(180px) 1fr 180px 100px 240px
+  grid-template-areas: "name label user provider time" "host host user provider time"
```
We have to do some fine-tuning at the grid I think. On smaller screens the timestamp overflows quite easily:
<img width="615" alt="Bildschirmfoto 2023-10-04 um 09 17 50" src="https://github.com/opf/openproject/assets/7457313/bd884ea1-d27f-418b-b960-efc8ac5a9b99">

Further, the label should probably not be the element that has `1fr` because it is not even visible in most cases. This leads to the name being cut off weirdly.
<img width="988" alt="Bildschirmfoto 2023-10-04 um 09 19 08" src="https://github.com/opf/openproject/assets/7457313/115f3079-d99b-43fe-af63-49b9752fe885">

HDinger reviewed (changes requested):
Hi @akabiru 

the page looks really great! 🥳  I have some remarks, but most of them are rather small. I can also help you with the needed CSS changes, if you want to. Further, I think we can discuss the need of the feature flag. If we remove that, we could get rid of some old code as well which would be nice :)

akabiru committed:
Revert "fix[#49932]: update timestamp format"

This reverts commit 5c61daf6a12ff3296ac5a5fd1009a32e1fd74190.

akabiru committed:
Revert "UI[#49932]: Resize primary buttons based on screen size"

This reverts commit bedbc14228313e34ada288c53c9b1609e6de460d.

akabiru committed:
fix[#49932]: Rename `label` grid area to `incomplete-label` as the more clear option

akabiru committed:
chore[#49932]: DRY up NewStorageButtonComponent options

akabiru committed:
fix[#49932]: Define font-wight in parent component

akabiru committed:
fix[#49932]: Use `hide: :sm` sys attribute instead for small screens

akabiru committed:
UI[#49932]: Merge name and incomplete-label into a single grid area

akabiru committed:
UI[#49932]: Differentiate header & row grid definitions

Co-authored-by: Henriette Darge <h.darge@openproject.com>

HDinger reviewed (approved):
Great job @akabiru 🥳 Thanks for all the iterations! The result was definitely worth it. Once the test selector is fixed, this can be merged 👍

akabiru committed:
Revert "Fix form submissions when mixing primer with named submit buttons"

This reverts commit 84cb848e418a7e599aa531388b17822581375a42.

akabiru committed:
fix[#49932]: Replace CSS classes in test with `data-test-selector`
